contracts/artifacts/DelegatedOptIn.approval.teal

```diff
@@ -0,0 +1,530 @@
+	int NoOp
+	==
+	txn ApplicationID
+	int 0
+	!=
+	&&
+	assert
+	byte 0x; dup
+	txna ApplicationArgs 1
+	btoi
+	txnas Assets
+	callsub updateAssetMBR
+	int 1
+	return
+
+updateAssetMBR:
+	proto 3 0
+
+	// ./contracts/delegated_optin_app.algo.ts:65
+	// preMbr = this.app.address.minBalance
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	acct_params_get AcctMinBalance
+	assert
+	frame_bury -2 // preMbr: uint64
+
+	// ./contracts/delegated_optin_app.algo.ts:67
+	// sendAssetTransfer({
+	itxn_begin
+	int axfer
+	itxn_field TypeEnum
+
+	// ./contracts/delegated_optin_app.algo.ts:68
+	// assetReceiver: this.app.address
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	itxn_field AssetReceiver
+
+	// ./contracts/delegated_optin_app.algo.ts:69
+	// xferAsset: asset
+	frame_dig -1 // asset: asset
+	itxn_field XferAsset
+
+	// ./contracts/delegated_optin_app.algo.ts:70
+	// assetAmount: 0
+	int 0
+	itxn_field AssetAmount
+
+	// ./contracts/delegated_optin_app.algo.ts:71
+	// fee: 0
+	int 0
+	itxn_field Fee
+	itxn_submit
+
+	// ./contracts/delegated_optin_app.algo.ts:74
+	// mbrDelta = preMbr - this.app.address.minBalance
+	frame_dig -2 // preMbr: uint64
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	acct_params_get AcctMinBalance
+	assert
+	-
+	frame_bury -3 // mbrDelta: uint64
+
+	// ./contracts/delegated_optin_app.algo.ts:76
+	// assert(mbrDelta !== this.assetMBR.get())
+	frame_dig -3 // mbrDelta: uint64
+	byte "assetMBR"
+	app_global_get
+	!=
+	assert
+
+	// ./contracts/delegated_optin_app.algo.ts:77
+	// this.assetMBR.set(mbrDelta)
+	byte "assetMBR"
+	frame_dig -3 // mbrDelta: uint64
+	app_global_put
+
+	// ./contracts/delegated_optin_app.algo.ts:79
+	// sendAssetTransfer({
+	itxn_begin
+	int axfer
+	itxn_field TypeEnum
+
+	// ./contracts/delegated_optin_app.algo.ts:80
+	// assetReceiver: this.app.address
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	itxn_field AssetReceiver
+
+	// ./contracts/delegated_optin_app.algo.ts:81
+	// xferAsset: asset
+	frame_dig -1 // asset: asset
+	itxn_field XferAsset
+
+	// ./contracts/delegated_optin_app.algo.ts:82
+	// assetAmount: 0
+	int 0
+	itxn_field AssetAmount
+
+	// ./contracts/delegated_optin_app.algo.ts:83
+	// fee: 0
+	int 0
+	itxn_field Fee
+
+	// ./contracts/delegated_optin_app.algo.ts:84
+	// assetCloseTo: this.app.address
+	txna Applications 0
+	app_params_get AppAddress
+	assert
+	itxn_field AssetCloseTo
+	itxn_submit
+	retsub
+
+abi_route_setOpenOptInSignature:
+	txn OnCompletion
+	int NoOp
+	==
+	txn ApplicationID
+	int 0
+	!=
+	&&
+	assert
+
+	// no dupn needed
+	txn GroupIndex
+	int 1
+	-
+	txna ApplicationArgs 2
+	txna ApplicationArgs 1
+	callsub setOpenOptInSignature
+	int 1
+	return
+
+setOpenOptInSignature:
+	proto 3 0
+
+	// ./contracts/delegated_optin_app.algo.ts:99
+	// assert(verifier.sender === this.sigVerificationAddress.get())
+	frame_dig -3 // verifier: txn
+	gtxns Sender
+	byte "sigVerificationAddress"
+	app_global_get
+	==
+	assert
```

Comment:
The current implementation allows *anyone* to upload the signature for a given address provided the signature is valid. If the app doesn't verify the lsigs, then it would have to implement logic that only allows the signer to put their signature in box storage. This would be counter-productive to one of the goals of reducing the onboarding friction of 0-ALGO accounts with an ASA.

Reply:
Oh, wait, maybe I should have looked before I spoke. Is it verifying the code, or just the signature on it?

In the 0-algo case, who is generating the signed lsig?

Reply:
Now that I see you're only putting the signature itself on-chain, ignore most of this.

But walk me through the 0-algo case. I don't get that. If someone is creating accounts, and therefore has keys, they could fund the account and perform the `set` transaction. Why are they willing and able to fund this thing, but not the account?

Reply:
In fact, I *think* they don't even have to fund the account. If they have keys, they could make the transaction to `set` the sig, by calling this app, without even funding the account, just creating the app call transaction and putting it into a group with their own account which pays the fees.

Reply:
Its just a matter of reducing the friction for getting the signature stored. The usefulness of this ARC is fairly dependent on the discoverability of signatures, so making it less clicks for the user to make their signature discoverable seems like it's worth the extra complexity of verification. In most scenarios (user interacting with an application), it's much easier for the app to programmatically set the sig box rather than having the user manually do it.

Reply:
> In most scenarios (user interacting with an application), it's much easier for the app to programmatically set the sig box rather than having the user manually do it.

If the user has to manually sign the lsig in the first place, I don't see why also having them sign and send a txn makes much of a difference.

And about the 0 algo case, it seems like users (those who want to opt-in others accounts) need to be aware of whether that account has 0 algos or not, since the MBR increase will be higher if they have to satisfy the base account MBR in addition to the asset MBR. This seems very annoying.

Reply:
If the app can do that, it must have the user's keys, right?  If so, it can set the signature by sending a transaction on the user's behalf. If it doesn't have the keys, how did it generate the signature it is supplying to the call?

It's hard for me to see how the signature can be generated, but a transaction can't be sent.

Reply:
The signature is still being created by the user. The difference is that the user uploading their own signatures means more clicks for them. They have to:

1. Create their account 
2. Sign the lsig
3. Share their address and signature
4. Wait for the app to respond with an atomic transaction that uses fee pooling from an account the app custodies
5. Sign the transaction

This is an obviously worse experience for the user because they have to sign the transaction and it's more complex for the app because they have to form the atomic transaction rather than simply calling the app with the signature.

Reply:
It's not obvious to me that's worse because I don't know which of these operations are supported in a natural way with any existing wallets.  Do any wallets support signing lsigs? or arbitrary hunks of data, supplied by a dapp?  If not, doesn't that kill both approaches?

Reply:
No they don't currently, but [ARC47](https://github.com/algorandfoundation/ARCs/pull/226) will enable wallets to sign known lsigs. A wallet *could* also abstract away the signing of the open opt-in delegation lsig and just have it be a radio button during account creation.